crates/core/src/auth/token_validation.rs

```diff
@@ -0,0 +1,613 @@
+    static ref REQUIRED_CLAIMS: Vec<&'static str> = vec!["sub", "iss"];
+}
+
+lazy_static! {
+    // Eventually we will want to add more required claims.
+    static ref GLOBAL_OIDC_VALIDATOR: Arc<CachingOidcTokenValidator> = Arc::new(CachingOidcTokenValidator::default());
+}
+
+// Just make some token website thing.
+
+// Get a minimal branch out that doesn't do any caching or fancy stuff.
+
+#[async_trait]
+impl TokenValidator for LocalTokenValidator {
+    async fn validate_token(&self, token: &str) -> Result<SpacetimeIdentityClaims2, TokenValidationError> {
+        // TODO: Make this stored in the struct so we don't need to keep creating it.
+        let mut validation = Validation::new(jsonwebtoken::Algorithm::ES256);
+        validation.algorithms = vec![
+            jsonwebtoken::Algorithm::ES256,
+            jsonwebtoken::Algorithm::RS256,
+            jsonwebtoken::Algorithm::HS256,
+        ];
+        validation.set_required_spec_claims(&REQUIRED_CLAIMS);
+        validation.set_issuer(&[self.issuer.clone()]);
+
+        // TODO: We should require a specific audience at some point.
+        validation.validate_aud = false;
+
+        let data = decode::<IncomingClaims>(token, &self.public_key, &validation)?;
+        let claims = data.claims;
+        if claims.issuer != self.issuer {
+            return Err(TokenValidationError::Other(anyhow::anyhow!(
+                "Issuer mismatch: got {:?}, expected {:?}",
+                claims.issuer,
+                self.issuer
+            )));
+        }
+        claims.try_into()
+    }
+}
+
+// Validates tokens by looking up public keys and caching them.
+pub struct CachingOidcTokenValidator {
+    cache: async_cache::AsyncCache<Arc<JwksValidator>, KeyFetcher>,
+}
+
+impl CachingOidcTokenValidator {
+    pub fn new(refresh_duration: Duration, expiry: Option<Duration>) -> Self {
+        let cache = async_cache::Options::new(refresh_duration, KeyFetcher)
+            .with_expire(expiry)
+            .build();
+        CachingOidcTokenValidator { cache }
+    }
+
+    pub fn default() -> Self {
+        Self::new(Duration::from_secs(300), Some(Duration::from_secs(7200)))
+    }
+}
+
+// Jwks fetcher for the async cache.
+struct KeyFetcher;
+
+impl async_cache::Fetcher<Arc<JwksValidator>> for KeyFetcher {
+    type Error = TokenValidationError;
+
+    async fn fetch(&self, key: FastStr) -> Result<Arc<JwksValidator>, Self::Error> {
+        // TODO: Make this stored in the struct so we don't need to keep creating it.
+        // let raw_issuer = get_raw_issuer(token)?;
+        let raw_issuer = key.to_string();
+        log::info!("Fetching key for issuer {}", raw_issuer.clone());
+        // TODO: Consider checking for trailing slashes or requiring a scheme.
+        let oidc_url = format!("{}/.well-known/openid-configuration", raw_issuer);
+        // TODO: log errors here.
+        let keys = Jwks::from_oidc_url(oidc_url).await?;
```

Comment:
Oh, convenient that there's a library for it.

Reply:
It won't be convenient if we switch `jwt` libraries, but it's simple to reimplement it.